Injected null in subclassed ContainerRequestFilter

**Describe the bug**
When extending a class which implements ContainerRequestFilter, the extended class gets @Inject'ed null in its fields.

**Expected behavior**
The injection is not null.

**Actual behavior**
```
lug 29, 2019 2:10:04 PM io.quarkus.undertow.runtime.QuarkusExceptionHandler handleThrowable
ERROR: Exception handling request b02cdeeb-cb97-478e-b2ee-dd0a7aad42cf-2 to /test/father
org.jboss.resteasy.spi.UnhandledException: java.lang.NullPointerException
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:249)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:57)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:175)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:28)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$8$1$1.call(UndertowDeploymentRecorder.java:483)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1429)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:32)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.lang.NullPointerException
	at org.example.filter.inheritance.FatherFilter.filter(FatherFilter.java:19)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:311)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:410)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:379)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invoke$1(ResourceMethodInvoker.java:353)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:353)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:477)
	... 49 more
```

**To Reproduce**
Steps to reproduce the behavior:
1. `git clone https://github.com/antoniomacri/quarkus-npe-filter-inheritance.git`
2. `mvn test`

**Environment (please complete the following information):**
 - Output of `java -version`: 1.8
 - Quarkus version or git rev:  0.19.1

**Additional context**
If I replace
```
public class SonFilter extends FatherFilter {
```
with
```
public class SonFilter implements ContainerRequestFilter {
```
it works.

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/AnnotationServletTestCase.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/AnnotationServletTestCase.java
@@ -15,7 +15,7 @@ public class AnnotationServletTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(TestServlet.class));
+                    .addClasses(TestServlet.class, TestServletSubclass.class, TestGreeter.class));
 
     @Test
     public void testServlet() {
@@ -24,4 +24,10 @@ public class AnnotationServletTestCase {
                 .body(is("test servlet"));
     }
 
+    @Test
+    public void testServletSubclass() {
+        RestAssured.when().get("/test-sub").then()
+                .statusCode(200)
+                .body(is("test servlet"));
+    }
 }

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/AnnotationWithWebXmlTestCase.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/AnnotationWithWebXmlTestCase.java
@@ -50,7 +50,7 @@ public class AnnotationWithWebXmlTestCase {
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                     .addClasses(WebXmlServlet.class, TestServlet.class,
                             WebXmlFilter.class, AnnotatedFilter.class,
-                            WebXmlListener.class, AnnotatedListener.class)
+                            WebXmlListener.class, AnnotatedListener.class, TestGreeter.class)
                     .addAsManifestResource(new StringAsset(WEB_XML), "web.xml"));
 
     @Test

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/ContextPathTestCase.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/ContextPathTestCase.java
@@ -18,7 +18,7 @@ public class ContextPathTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(TestServlet.class)
+                    .addClasses(TestServlet.class, TestGreeter.class)
                     .addAsResource(new StringAsset("quarkus.servlet.context-path=" + CONTEXT_PATH), "application.properties"));
 
     @Test

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/TestGreeter.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/TestGreeter.java
@@ -1,0 +1,11 @@
+package io.quarkus.undertow.test;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class TestGreeter {
+
+    public String message() {
+        return "test servlet";
+    }
+}

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/TestServlet.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/TestServlet.java
@@ -2,6 +2,7 @@ package io.quarkus.undertow.test;
 
 import java.io.IOException;
 
+import javax.inject.Inject;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
@@ -10,8 +11,11 @@ import javax.servlet.http.HttpServletResponse;
 @WebServlet(urlPatterns = "/test")
 public class TestServlet extends HttpServlet {
 
+    @Inject
+    TestGreeter testGreeter;
+
     @Override
     protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
-        resp.getWriter().write("test servlet");
+        resp.getWriter().write(testGreeter.message());
     }
 }

--- a/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/TestServletSubclass.java
+++ b/extensions/undertow/deployment/src/test/java/io/quarkus/undertow/test/TestServletSubclass.java
@@ -1,0 +1,8 @@
+package io.quarkus.undertow.test;
+
+import javax.servlet.annotation.WebServlet;
+
+@WebServlet(urlPatterns = "/test-sub")
+public class TestServletSubclass extends TestServlet {
+
+}

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
@@ -75,6 +75,11 @@ public interface ArcContainer {
     /**
      * Returns a supplier that can be used to create new instances, or null if no matching bean can be found.
      *
+     * Note that if there are multiple sub classes of the given type this will return the exact match. This means
+     * that this can be used to directly instantiate superclasses of other beans without causing problems.
+     *
+     * see https://github.com/quarkusio/quarkus/issues/3369
+     *
      * @param type
      * @param qualifiers
      * @param <T>

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java
@@ -213,7 +213,24 @@ public class ArcContainerImpl implements ArcContainer {
     public <T> Supplier<InstanceHandle<T>> instanceSupplier(Class<T> type, Annotation... qualifiers) {
         requireRunning();
 
-        InjectableBean<T> bean = getBean(type, qualifiers);
+        if (qualifiers == null || qualifiers.length == 0) {
+            qualifiers = new Annotation[] { Default.Literal.INSTANCE };
+        }
+        Set<InjectableBean<?>> resolvedBeans = resolved.getValue(new Resolvable(type, qualifiers));
+        Set<InjectableBean<?>> filteredBean = resolvedBeans;
+        if (resolvedBeans.size() > 1) {
+            //if there are multiple beans we look for an exact match
+            //this method is only called with the exact type required
+            //so ignoring subclasses is the correct behaviour
+            filteredBean = new HashSet<>();
+            for (InjectableBean<?> i : resolvedBeans) {
+                if (i.getBeanClass().equals(type)) {
+                    filteredBean.add(i);
+                }
+            }
+        }
+        InjectableBean<T> bean = filteredBean.isEmpty() || filteredBean.size() > 1 ? null
+                : (InjectableBean<T>) filteredBean.iterator().next();
         if (bean == null) {
             return null;
         }

